src/com/amazon/corretto/crypto/provider/HpkeCipher.java

```diff
@@ -0,0 +1,309 @@
+    }
+    if (((EvpHpkeKey) key).getSpec() != params) {
+      throw new InvalidKeyException("Spec of key does not match the params provided");
+    }
+    if ((opmode == Cipher.WRAP_MODE) && !(key instanceof EvpHpkePublicKey)) {
+      throw new IllegalStateException("Need PublicKey to wrap");
+    }
+    if ((opmode == Cipher.UNWRAP_MODE) && !(key instanceof EvpHpkePrivateKey)) {
+      throw new IllegalStateException("Need PrivateKey to unwrap");
+    }
+    synchronized (lock_) {
+      params_ = (HpkeParameterSpec) params;
+      javaCipherMode_ = opmode;
+      key_ = opmode == Cipher.WRAP_MODE ? (EvpHpkePublicKey) key : (EvpHpkePrivateKey) key;
+    }
+  }
+
+  /**
+   * Performs HPKE single-shot encryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return concatenation of KEM encapsulated key and encrypted ciphertext
+   */
+  @Override
+  protected byte[] engineWrap(Key key) throws IllegalBlockSizeException, InvalidKeyException {
+    if (javaCipherMode_ != Cipher.WRAP_MODE) {
+      throw new IllegalStateException("Cipher must be in WRAP_MODE");
+    }
+    if ((key_ == null) || !(key_ instanceof EvpHpkePublicKey)) {
+      throw new IllegalStateException("PublicKey should be set before wrapping.");
+    }
+    try {
+      final byte[] encoded = Utils.encodeForWrapping(provider_, key);
+      return engineDoFinal(encoded, 0, encoded.length);
+    } catch (final BadPaddingException e) {
+      throw new InvalidKeyException("Failed to wrap key", e);
+    }
+  }
+
+  /**
+   * Performs HPKE single-shot decryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return decrypted plaintext
+   */
+  @Override
+  protected Key engineUnwrap(byte[] wrappedKey, String wrappedKeyAlgorithm, int wrappedKeyType)
+      throws InvalidKeyException, NoSuchAlgorithmException {
+    if (javaCipherMode_ != Cipher.UNWRAP_MODE) {
+      throw new IllegalStateException("Cipher must be in UNWRAP_MODE");
+    }
+    if ((key_ == null) || !(key_ instanceof EvpHpkePrivateKey)) {
+      throw new IllegalStateException("PrivateKey should be set before unwrapping.");
+    }
+    try {
+      final byte[] unwrappedKey = engineDoFinal(wrappedKey, 0, wrappedKey.length);
+      return Utils.buildUnwrappedKey(provider_, unwrappedKey, wrappedKeyAlgorithm, wrappedKeyType);
+    } catch (final BadPaddingException | IllegalBlockSizeException | InvalidKeySpecException e) {
+      throw new InvalidKeyException("Failed to unwrap key", e);
+    }
+  }
+
+  /**
+   * Performs HPKE single-shot encryption or decryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return number of bytes written to output
+   */
+  @Override
+  protected int engineDoFinal(
+      byte[] input, int inputOffset, int inputLen, byte[] output, int outputOffset)
+      throws ShortBufferException {
+
+    if (input == output) {
+      // TODO: is it okay if they are non-overlapping, support that case if necessary.
```

Comment:
Check out the AES-GCM implementation, we go through a lot of trouble to support this. I'd rank this below all the other stuff like encrypt/decrypt and AAD though.

Reply:
Added support for overlapping buffers.